### ISSUE TITLE
GA for Feature Selection - Issues with setup?

### ISSUE THREAD

**Issue by GinoWoz1** [DEAP/deap#308]
Hello,

First of all, thanks so much for creating this package, so easy to use! I have already used it a bunch indirectly through TPOT.

Currently I am trying to use deap for feature selection and running into some issues that I hope you can help me with **(sorry for the formatting, not sure why this post is formatted weird)**.

I am using the boston house data set and created some custom house groupings + added polynomial features and interaction terms.  Please see my setup below with the "issue" at the end. 

**With 74 columns i am using the toolbox below:**

# Create Individual
creator.create("FitnessMax", base.Fitness, weights=(1.0,))
creator.create("Individual", list, fitness=creator.FitnessMax)

# Create Toolbox
toolbox = base.Toolbox()
toolbox.register("attr_bool", random.randint, 0, 1)
toolbox.register("individual", tools.initRepeat, creator.Individual, toolbox.attr_bool, len(X_train_poly.columns) - 1)
toolbox.register("population", tools.initRepeat, list, toolbox.individual)

# Continue filling toolbox...
toolbox.register("evaluate", getFitness, X_train=X_train_poly, y_train=y_train)
toolbox.register("mate", tools.cxTwoPoint)
toolbox.register("mutate", tools.mutFlipBit, indpb=0.10)
toolbox.register("select", tools.selNSGA2)

# Here is my scorer

def getFitness(individual, X_train, y_train):

	# Parse our feature columns that we don't use
	# Apply one hot encoding to the features
	# Apply logistic regression on the data, and calculate accuracy
	elnet_final = ElasticNet(alpha=0,l1_ratio=0.0,tol=1e-05)
	cross_val = cross_val_score(elnet_final,X_train_poly,y_train,scoring=rmse_cv,cv=rkfold)
	score = np.mean(cross_val)

	# Return calculated accuracy as fitness
	return (score,)

# Here is my function to initiate

def getHof():
	# Initialize variables to use eaSimple
	numPop = 200
	numGen = 100
	pop = toolbox.population(n=numPop)
	hof = tools.HallOfFame(numPop * numGen)
	stats = tools.Statistics(lambda ind: ind.fitness.values)
	stats.register("avg", numpy.mean)
	stats.register("std", numpy.std)
	stats.register("min", numpy.min)
	stats.register("max", numpy.max)

	# Launch genetic algorithm
	pop, log = algorithms.eaSimple(pop, toolbox, cxpb=0.9, mutpb=0.1, ngen=numGen, stats=stats, halloffame=hof, verbose=True)

	# Return the hall of fame
	return hof

However, when I start the GA run, my "max" score doesnt improve. First of all, my score is returning a negative (similar to some sklearn scores) so I use the FitnessMax for my fitness scoring mechanism. With a population size of 200 I am using the two-point crossover based on some papers I have been reading.

Am I doing something wrong? Should I be using Uniform crossover? I was able to get a score of -3.4 using SelectKBest with the top 50 variables (and cross_val_score) and I am not getting close to that.

![image](https://user-images.githubusercontent.com/17535345/47383504-acba8980-d6b9-11e8-9beb-229ce27d4c41.png)

Sincerely,
G

**Comment by cmd-ntrf** [comment]
You are currently using `selNSGA2` as a selection operator. NSGA2 is a multi-objective selection algorithm.

Your fitness function returns a single value and your FitnessMax's weight have also a single value. I therefore conclude that you are trying to do single-objective optimization. 

Change your selection function to a single objective algorithm, like `selTournament`.

**Comment by GinoWoz1** [comment]
Thank you so much. I appreciate it. 

In terms of feature selection I am using 73 variables which comes out to an absurd amount of possible combinations. Are there any custom statistics that one can build to show the homogeneity of each generation? At one point last week I was getting below the score below with a tournament size of 128 (out of 730 population) using Mu + Lambda method. Lately I havent gotten close to it which made me assume that I was reaching a local optimum and not incorporating other combinations (using a crossover rate of 90% and a mutation rate of 10%). To try and improve the diversity I changed the crossover rate to 60 % and the mutation rate to 40% (to flip bits).  

Am I approaching this wrong? 

What I have tested:
OnePoint crossover (second worst performance)
TwoPoint crossover (worst performance)
Uniform crossover (best performance)

Tournament size:
128, 64, 32, 16, 4 and 2 - pretty much getting stuck at local optimum.

Crossover/mutation rate:
90 % crossover and 10% mutation (got suck at local minimum)
80% crossover and 10% mutation (improved slightly)
60 % crossover and 10% mutation (current testing)

Current run (mu + lambda, 730 pop size 730 mu and lambda, 73 columns, 60% crossover rate, 40% mutation rate, tournament size =4)

![image](https://user-images.githubusercontent.com/17535345/47607792-7e9ab980-d9d9-11e8-8ec2-97ea382865ef.png)

**Comment by fmder** [comment]
What kind of variables do you have? For continuous bounded variables
polynomials crossover often outperform point crossover.

**Comment by GinoWoz1** [comment]
Thanks, My variables include 1 binary variable with 1 nominal variable and then 71 continuous variables. 11 of the continuous variables are in the raw, the other 60 are either polynomial interactions or polynomial terms to the 2nd order. 

I do some eda in the workbook below although you do not see my polynomial continuous variables in the eda.

http://nbviewer.jupyter.org/github/GinoWoz1/Learnings/blob/master/Boston_Dataset.ipynb

**Comment by fmder** [comment]
Sorry, I missed that you are doing feature selection, so you are evolving a
binary mask. Forget about polynomial crossover.

**Comment by GinoWoz1** [comment]
when you set the mutflipbit rate and the mutation rate, is the 20% evaluated against every binary value?

Also with select tournament size, is it selecting for example 4 members for EVERY crossover ; in this case with MU + Lambda I am assuming it is doing the select tournament size for every crossover. Sorry for the stupid questions.

Sincerely,
G

**Comment by GinoWoz1** [comment]
Reading this kdnuggets aritcle, seems I maybe need to approach it from a dual optimization perspective using the loss function and # of variables. Could be wrong.

https://www.kdnuggets.com/2017/12/rapidminer-multi-objective-optimization-feature-selection.html

**Comment by GinoWoz1** [comment]
Hello @fmder , sorry for the bother, but how can you print out 2 statistics with multi objective optimization? 

I am setting up per the below but it is not printing out like the default single statistics setup. I spent about 5-6 s hours on this but couldnt get it to work =/ Searched the examples provided but still the same result.

**fitness**
def getFitness(individual, X_train_poly, y_train):
    # Parse our feature columns that we don't use
    # Apply one hot encoding to the features
    # Apply logistic regression on the data, and calculate accuracy
    elnet_final = ElasticNet(alpha=0,l1_ratio=0.0,tol=1e-05)
    cross_val = cross_val_score(elnet_final,X_train_poly,y_train,scoring=rmse_cv,cv=rkfold)
    score = np.mean(cross_val)
       
    feature_count =  0
    for i in range(len(individual)):
        binary = individual[i]
        if binary == 1:
            feature_count+=1
    
    # Return calculated accuracy as fitness
    return (score,feature_count)

**HOF call function**
def getHof():
    # Initialize variables to use eaSimple
    numPop = 730
    numGen = 100
    pop = toolbox.population(n=numPop)
    hof = tools.HallOfFame(numPop * numGen)    
    stats_loss = tools.Statistics(key=lambda ind: ind.fitness.values[0])
    stats_size = tools.Statistics(key=lambda ind: ind.fitness.values[1])
    mstats = tools.MultiStatistics(fitness=stats_loss, size=stats_size)
    mstats.register("avgLoss", np.mean)
    mstats.register("stdLoss", np.std)
    mstats.register("minLoss", np.min)
    mstats.register("maxLoss", np.max)

        
    # Launch genetic algorithm
    pop, log = algorithms.eaSimple(pop, toolbox, cxpb=0.9, mutpb=0.1, ngen=numGen, stats=mstats, halloffame=hof, verbose=True)
    
    # Return the hall of fame
    return hof
